Adjust GC Timer a bit with heuristics to avoid running it while critical tasks are waiting

Constellation opened the pull request:
#### 5c9eac47597dd0447d181e32ef2436cfce67791c
<pre>
Adjust GC Timer a bit with heuristics to avoid running it while critical tasks are waiting
<a href="https://bugs.webkit.org/show_bug.cgi?id=265055">https://bugs.webkit.org/show_bug.cgi?id=265055</a>
<a href="https://rdar.apple.com/118574205">rdar://118574205</a>

Reviewed by Wenson Hsieh and Justin Michaud.

We found that GC timer fires randomly and it runs GC at random timing. We should avoid
running it when there is critical tasks are waiting. This patch starts with a naive simple
approach which just defers this GC invocation with some threshold when there is a waiting task.
We are not directly using OpportunisticTaskScheduler since it is active only when a page is visible
and active, but we would like to run this GC for background page too. Eventually we would like
to unify both into one global mechanism but we put it as a future work.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/heap/EdenGCActivityCallback.cpp:
(JSC::EdenGCActivityCallback::EdenGCActivityCallback):
(JSC::EdenGCActivityCallback::doCollection):
* Source/JavaScriptCore/heap/EdenGCActivityCallback.h:
(JSC::EdenGCActivityCallback::tryCreate):
(): Deleted.
(JSC::GCActivityCallback::tryCreateEdenTimer): Deleted.
* Source/JavaScriptCore/heap/FullGCActivityCallback.cpp:
(JSC::FullGCActivityCallback::FullGCActivityCallback):
(JSC::FullGCActivityCallback::doCollection):
* Source/JavaScriptCore/heap/FullGCActivityCallback.h:
(JSC::FullGCActivityCallback::tryCreate):
(): Deleted.
(JSC::GCActivityCallback::tryCreateFullTimer): Deleted.
* Source/JavaScriptCore/heap/GCActivityCallback.cpp:
(JSC::GCActivityCallback::GCActivityCallback):
(JSC::GCActivityCallback::scheduleTimer):
* Source/JavaScriptCore/heap/GCActivityCallback.h:
(JSC::GCActivityCallback::didGCRecently const):
(JSC::GCActivityCallback::setDidGCRecently):
(JSC::GCActivityCallback::GCActivityCallback):
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::Heap):
(JSC::Heap::collect):
(JSC::Heap::updateAllocationLimits):
(JSC::Heap::collectNowFullIfNotDoneRecently):
(JSC::Heap::setFullActivityCallback):
(JSC::Heap::setEdenActivityCallback):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/heap/MarkedSpace.cpp:
(JSC::MarkedSpace::beginMarking):
* Source/JavaScriptCore/heap/MarkedSpace.h:
(JSC::MarkedSpace::edenVersion const):
* Source/JavaScriptCore/runtime/DeferredWorkTimer.h:
* Source/JavaScriptCore/runtime/JSRunLoopTimer.h:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::performOpportunisticallyScheduledTasks):
* Source/WebCore/bindings/js/CommonVM.cpp:
(WebCore::commonVMSlow):
* Source/WebCore/page/OpportunisticTaskScheduler.cpp:
(WebCore::OpportunisticTaskScheduler::runLoopObserverFired):
(WebCore::isBusyForTimerBasedGC):
(WebCore::OpportunisticTaskScheduler::FullGCActivityCallback::doCollection):
(WebCore::OpportunisticTaskScheduler::EdenGCActivityCallback::doCollection):
* Source/WebCore/page/OpportunisticTaskScheduler.h:
(WebCore::OpportunisticTaskScheduler::create): Deleted.
(WebCore::OpportunisticTaskScheduler::willQueueIdleCallback): Deleted.
(WebCore::OpportunisticTaskScheduler::hasImminentlyScheduledWork const): Deleted.

Canonical link: <a href="https://commits.webkit.org/270919@main">https://commits.webkit.org/270919@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b3d0b0bb0ea6c7a6eaa0d79856c5b7b76b9ccb95

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26815 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5431 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28051 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29026 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24509 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27261 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7265 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2822 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27077 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4244 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23013 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3723 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/3770 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24010 "Passed tests") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29506 "Hash b3d0b0bb for PR 20688 does not build (failure)") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/23365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/24454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24411 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/29506 "Hash b3d0b0bb for PR 20688 does not build (failure)") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/26002 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/3803 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/1998 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/29506 "Hash b3d0b0bb for PR 20688 does not build (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5256 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/33455 "Built successfully") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4270 "Build is in progress. Recent messages:") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7232 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3472 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4165 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/page/OpportunisticTaskScheduler.cpp:
```diff
@@ -98,10 +106,12 @@ void OpportunisticTaskScheduler::runLoopObserverFired()
         if (m_runloopCountAfterBeingScheduled > minimumRunloopCountWhenScheduledWorkIsImminent)
             return true;
 
+        dataLogLnIf(verbose, "GAVEUP: task does not get scheduled ", remainingTime, " ", hasImminentlyScheduledWork(), " ", page->preferredRenderingUpdateInterval(), " ", m_runloopCountAfterBeingScheduled);
```
Interesting! We don't generally use dataLogs in WebCore and above, but it's probably fine for this use, given that this code is closely tied to the GC lifecycle.

I wonder if we could add something like `"[OPPORTUNISTIC TASK]"` to the beginning of all these datalog message, so that it's easier to filter by that when debugging/analyzing the OpTask lifecycle.

Constellation replied:
`"[OPPORTUNISTIC TASK]"` sounds good!

whsieh reviewed (approved):
The WebCore and VM changes seem reasonable to me, but it might be good to have a JSC reviewer take a peek at the other JSC changes too!

justinmichaud reviewed (approved):
r=me too, LGTM

Constellation committed:
Adjust GC Timer a bit with heuristics to avoid running it while critical tasks are waiting
https://bugs.webkit.org/show_bug.cgi?id=265055
rdar://118574205

Reviewed by Wenson Hsieh and Justin Michaud.

We found that GC timer fires randomly and it runs GC at random timing. We should avoid
running it when there is critical tasks are waiting. This patch starts with a naive simple
approach which just defers this GC invocation with some threshold when there is a waiting task.
We are not directly using OpportunisticTaskScheduler since it is active only when a page is visible
and active, but we would like to run this GC for background page too. Eventually we would like
to unify both into one global mechanism but we put it as a future work.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/heap/EdenGCActivityCallback.cpp:
(JSC::EdenGCActivityCallback::EdenGCActivityCallback):
(JSC::EdenGCActivityCallback::doCollection):
* Source/JavaScriptCore/heap/EdenGCActivityCallback.h:
(JSC::EdenGCActivityCallback::tryCreate):
(): Deleted.
(JSC::GCActivityCallback::tryCreateEdenTimer): Deleted.
* Source/JavaScriptCore/heap/FullGCActivityCallback.cpp:
(JSC::FullGCActivityCallback::FullGCActivityCallback):
(JSC::FullGCActivityCallback::doCollection):
* Source/JavaScriptCore/heap/FullGCActivityCallback.h:
(JSC::FullGCActivityCallback::tryCreate):
(): Deleted.
(JSC::GCActivityCallback::tryCreateFullTimer): Deleted.
* Source/JavaScriptCore/heap/GCActivityCallback.cpp:
(JSC::GCActivityCallback::GCActivityCallback):
(JSC::GCActivityCallback::scheduleTimer):
* Source/JavaScriptCore/heap/GCActivityCallback.h:
(JSC::GCActivityCallback::didGCRecently const):
(JSC::GCActivityCallback::setDidGCRecently):
(JSC::GCActivityCallback::GCActivityCallback):
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::Heap):
(JSC::Heap::collect):
(JSC::Heap::updateAllocationLimits):
(JSC::Heap::collectNowFullIfNotDoneRecently):
(JSC::Heap::setFullActivityCallback):
(JSC::Heap::setEdenActivityCallback):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/heap/MarkedSpace.cpp:
(JSC::MarkedSpace::beginMarking):
* Source/JavaScriptCore/heap/MarkedSpace.h:
(JSC::MarkedSpace::edenVersion const):
* Source/JavaScriptCore/runtime/DeferredWorkTimer.h:
* Source/JavaScriptCore/runtime/JSRunLoopTimer.h:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::performOpportunisticallyScheduledTasks):
* Source/WebCore/bindings/js/CommonVM.cpp:
(WebCore::commonVMSlow):
* Source/WebCore/page/OpportunisticTaskScheduler.cpp:
(WebCore::OpportunisticTaskScheduler::runLoopObserverFired):
(WebCore::isBusyForTimerBasedGC):
(WebCore::OpportunisticTaskScheduler::FullGCActivityCallback::doCollection):
(WebCore::OpportunisticTaskScheduler::EdenGCActivityCallback::doCollection):
* Source/WebCore/page/OpportunisticTaskScheduler.h:
(WebCore::OpportunisticTaskScheduler::create): Deleted.
(WebCore::OpportunisticTaskScheduler::willQueueIdleCallback): Deleted.
(WebCore::OpportunisticTaskScheduler::hasImminentlyScheduledWork const): Deleted.

Canonical link: https://commits.webkit.org/270919@main
